Pod/Classes/PINRemoteImageManager.m

```diff
@@ -1165,23 +1176,35 @@ - (NSUUID *)downloadImageWithURLs:(NSArray *)urls
             desiredImageURLIdx = ceilf((currentBytesPerSecond - lowQualityQPSThreshold) / ((highQualityQPSThreshold - lowQualityQPSThreshold) / (float)(urls.count - 2)));
         }
         
-        NSURL *downloadURL;
+        NSUInteger downloadIdx;
         //if the highest quality already downloaded is less than what currentBPS would dictate and shouldUpgrade is
         //set, download the new higher quality image. If no image has been cached, download the image dictated by
         //current bps
         if ((highestQualityDownloadedIdx < desiredImageURLIdx && shouldUpgradeLowQualityImages) || highestQualityDownloadedIdx == -1) {
-            downloadURL = [urls objectAtIndex:desiredImageURLIdx];
+            downloadIdx = desiredImageURLIdx;
         } else {
-            downloadURL = [urls objectAtIndex:highestQualityDownloadedIdx];
+            downloadIdx = highestQualityDownloadedIdx;
         }
         
+        NSURL *downloadURL = [urls objectAtIndex:downloadIdx];
+        
         [strongSelf downloadImageWithURL:downloadURL
                                  options:options
                                 priority:PINRemoteImageManagerPriorityMedium
                             processorKey:nil
                                processor:nil
                                 progress:progress
-                              completion:completion
+                              completion:^(PINRemoteImageManagerResult *result) {
+                                  typeof(self) strongSelf = weakSelf;
+                                  //clean out any lower quality images from the cache
+                                  for (NSInteger idx = downloadIdx - 1; idx >= 0; idx--) {
+                                      [[strongSelf cache] removeObjectForKey:[strongSelf cacheKeyForURL:[urls objectAtIndex:idx] processorKey:nil]];
+                                  }
```

Comment:
This looks fine but makes me think we should consider a `-removeObjectsForKeys:(NSArray *)` method in PINCache that can do this work in a single locked batch operation.